iambic/plugins/v0_1_0/aws/organizations/scp/models.py

```diff
@@ -0,0 +1,582 @@
+from __future__ import annotations
+
+import asyncio
+from enum import Enum
+from itertools import chain
+from typing import TYPE_CHECKING, Callable, List, Optional, TypedDict, Union
+
+from pydantic import BaseModel as PydanticBaseModel
+from pydantic import Field, constr, validator
+
+from iambic.core.context import ctx
+from iambic.core.iambic_enum import Command
+from iambic.core.logger import log
+from iambic.core.models import (
+    AccountChangeDetails,
+    BaseModel,
+    ExpiryModel,
+    ProposedChange,
+    ProposedChangeType,
+)
+from iambic.core.utils import normalize_dict_keys, plugin_apply_wrapper
+from iambic.plugins.v0_1_0.aws.models import (
+    AccessModel,
+    AWSAccount,
+    AWSTemplate,
+    Description,
+    Tag,
+)
+from iambic.plugins.v0_1_0.aws.organizations.scp.utils import (
+    apply_update_policy,
+    apply_update_policy_tags,
+    apply_update_policy_targets,
+    create_policy,
+    delete_policy,
+    get_policy,
+    service_control_policy_is_enabled,
+)
+
+if TYPE_CHECKING:
+    from iambic.plugins.v0_1_0.aws.iam.models import Path
+    from iambic.plugins.v0_1_0.aws.iambic_plugin import AWSConfig
+    from iambic.plugins.v0_1_0.aws.models import AWSOrganization
+
+
+AWS_SCP_POLICY_TEMPLATE = "NOQ::AWS::Organizations::SCP"
+
+
+class StatementEffect(str, Enum):
+    ALLOW = "Allow"
+    DENY = "Deny"
+
+
+class ServiceControlPolicyTargetItemType(str, Enum):
+    ACCOUNT = "ACCOUNT"
+    ORGANIZATIONAL_UNIT = "ORGANIZATIONAL_UNIT"
+    ROOT = "ROOT"
+
+
+class ServiceControlPolicyItemType(str, Enum):
+    SERVICE_CONTROL_POLICY = "SERVICE_CONTROL_POLICY"
+    TAG_POLICY = "TAG_POLICY"
+    BACKUP_POLICY = "BACKUP_POLICY"
+    AISERVICES_OPT_OUT_POLICY = "AISERVICES_OPT_OUT_POLICY"
+
+
+class OutputModels(PydanticBaseModel):
+    """Output models from AWS Organizations client"""
+
+    class Config:
+        use_enum_values = True
+
+
+class PolicyStatement(AccessModel, ExpiryModel):
```

Comment:
Ditto on thinking that we cover this in the core IAM models already and should reuse

Reply:
I think there is a situation here. Policy statements in IAM are similar but not igual than SCP policies statements. 
Unsupported elements at SCP Policy statements:
    - Principal
    - NotPrincipal
    - NotResource
Is not possible to "remove" fields, but I am not sure if it is correct to inherith one from another, as well I don't know if other policies in other services needs other restrictions. 

What I'm thinking as a solution, is to have a BasePolicyStatement at aws/models.py and each service has its own implementation, besides some of them could have directly a `pass`on the class body. 


```suggestion
class PolicyStatement(AccessModel, ExpiryModel):
```

Reply:
Yeah the base policy model makes sense. The `Principal` element is only supported in AWS Resource / Trust Policies. Not in IAM policy statements either.

principal policy statements in inline policies, managed policies (and permission boundaries), service control policies, and session policies are the same. (The support action, notaction, effect, resource, notresource, condition, but they do not support principal)

Role trust policies, and resource policies support PRINCIPAL, action, effect, resource, and condition.

With that in mind, I believe we're using `PolicyStatement` for both principal policies, AND role trust policies. At the bare minimum, I think we can do the following (But it can be later):

1. Use PolicyStatement for SCPs as well
2. Create a base policy model per your recommendation
3. Create a PolicyStatement type for assume_role (trust) Policy and replace the trust policy model to use it

Reply:
Task created https://noqdev.atlassian.net/browse/EN-2130